pkg/operator/clustermemberremovalcontroller/clustermemberremovalcontroller.go

```diff
@@ -0,0 +1,253 @@
+	etcdClient                        etcdcli.EtcdClient
+	masterNodeLister                  corev1listers.NodeLister
+	masterMachineLister               machinelistersv1beta1.MachineLister
+	networkLister                     configv1listers.NetworkLister
+	configMapListerForTargetNamespace corev1listers.ConfigMapNamespaceLister
+	hasNetworkInformerSyncedFn        func() bool
+	// isMachineAPIFunctionalFn determines if the precondition for this controller is met,
+	// this controller can be run only on a cluster that exposes a functional Machine API
+	isMachineAPIFunctionalFn func() (bool, error)
+
+	masterMachineSelector labels.Selector
+	masterNodeSelector    labels.Selector
+}
+
+// NewClusterMemberRemovalController removes an etcd member if the machine and a node for the etcd member is gone and the machine-api is active
+//
+// Note:
+//  since this controller needs to reconcile only master nodes and machine objects
+//  make sure nodeInformer and machineInformer contain only filtered data
+//  otherwise it might be expensive to react to every node update in larger installations
+func NewClusterMemberRemovalController(
+	etcdClient etcdcli.EtcdClient,
+	isMachineAPIFunctional func() (bool, error),
+	masterMachineSelector labels.Selector, masterNodeSelector labels.Selector,
+	kubeInformersForNamespaces operatorv1helpers.KubeInformersForNamespaces,
+	masterNodeInformer cache.SharedInformer,
+	masterMachineInformer cache.SharedInformer,
+	masterNodeLister corev1listers.NodeLister,
+	masterMachineLister machinelistersv1beta1.MachineLister,
+	hasNetworkInformerSyncedFn func() bool,
+	networkLister configv1listers.NetworkLister,
+	eventRecorder events.Recorder,
+) factory.Controller {
+	c := &ClusterMemberRemovalController{
+		etcdClient:                        etcdClient,
+		isMachineAPIFunctionalFn:          isMachineAPIFunctional,
+		masterMachineSelector:             masterMachineSelector,
+		masterNodeSelector:                masterNodeSelector,
+		masterNodeLister:                  masterNodeLister,
+		masterMachineLister:               masterMachineLister,
+		networkLister:                     networkLister,
+		configMapListerForTargetNamespace: kubeInformersForNamespaces.InformersFor(operatorclient.TargetNamespace).Core().V1().ConfigMaps().Lister().ConfigMaps(operatorclient.TargetNamespace),
+		hasNetworkInformerSyncedFn:        hasNetworkInformerSyncedFn,
+	}
+	return factory.New().
+		WithSync(c.sync).
+		ResyncEvery(33*time.Minute). // make it slow since nodes are updated every few minutes
+		WithInformers(
+			masterNodeInformer,
+			masterMachineInformer,
+			kubeInformersForNamespaces.InformersFor(operatorclient.TargetNamespace).Core().V1().ConfigMaps().Informer(),
+		).ToController("ClusterMemberRemovalController", eventRecorder.WithComponentSuffix("cluster-member-removal-controller"))
+}
+
+func (c *ClusterMemberRemovalController) sync(ctx context.Context, _ factory.SyncContext) error {
+	if !c.hasNetworkInformerSyncedFn() {
```

Comment:
normally you add this to "must be required to sync before starting the consumer"  It's a cacheSync list somewhere.

Reply:
yes, this is true when an informer is added via the `WithInformers` method.

However, in this case, since we are not interested in reacting to changes made to this informer (no call to `WithInformers`) we need to make sure it has synced explicitly.

Reply:
oh, it looks like we could use `WithBareInformers` - awesome!